src/mds/QuiesceAgent.h

```diff
@@ -0,0 +1,249 @@
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2023 IBM, Red Hat
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License db_version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+#pragma once
+#include "mds/QuiesceDb.h"
+#include <functional>
+#include <optional>
+#include <map>
+#include <mutex>
+#include <thread>
+
+class QuiesceAgent {
+  public:
+    struct ControlInterface {
+      QuiesceInterface::RequestSubmit submit_request;
+      QuiesceInterface::RequestCancel cancel_request;
+      QuiesceInterface::AgentAck agent_ack;
+      // TODO: do we need a "cancel all"?
+    };
+
+    QuiesceAgent(const ControlInterface& quiesce_control)
+        : quiesce_control(quiesce_control)
+        , stop_agent_thread(false)
+        , agent_thread(this) {
+      agent_thread.create("quiesce.agt");
+    };
+
+    ~QuiesceAgent() {
+      agent_thread.kill(SIGTERM);
```

Comment:
Killing a thread is a code smell. Just set the shutdown flag and join. If it blocks forever, so be it.

Reply:
Good catch! That thread could be holding a mutex and it seems like killing a thread doesn't release the mutex